### PR TITLE
New familyId argument

### PR DESCRIPTION
**Pull request by jose-oc**
- New familyId argument for accounts with more than one member
- Dependecies upgraded
- Logs
- Output messages rather than exceptions on the console
- Dockerfile rewritten to
  - Save build time when the app is changed (libraries are built before copying the app)
  - Accept arguments to run the app directly without Makefile

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,10 +1,9 @@
-FROM python:latest
+FROM python:3.11
```
Comment (pablobuenaposada) [inline Dockerfile:1]: I can buy it, no prob

File: Dockerfile [hunk 2]
```diff
@@ -1,10 +1,9 @@
...
-COPY Makefile requirements.txt requirements-tests.txt setup.cfg /usr/src/app/
+COPY requirements.txt ./
+RUN pip install --no-cache-dir -r requirements.txt
 
-RUN make venv
-
-CMD ["make", "run"]
\ No newline at end of file
+COPY src/ /usr/src/app/
+ENTRYPOINT [ "python", "./main.py" ]
```
Comment (pablobuenaposada) [inline Dockerfile:9]: I don't see the point of no having a centralize file with all the available "commands" you can do

File: requirements.txt
Comment (pablobuenaposada) [inline requirements.txt:1]: agree

File: src/client.py [hunk 3]
```diff
@@ -25,6 +27,10 @@ def __init__(self, email: str, password: str, box_id: int, box_name: str):
         self.box_id = box_id
         self.box_name = box_name
 
+        self.logger = logging.getLogger(__name__)
```
Comment (pablobuenaposada) [inline src/client.py:30]: what's the point of logging in this simple project? I mean if there's something wrong the exceptions right now are pretty self explanatory

File: src/client.py [hunk 4]
```diff
@@ -45,32 +51,46 @@ def _login(email: str, password: str):
...
-    def book_class(self, target_day: datetime, class_id: str) -> bool:
+    def book_class(self, target_day: datetime, class_id: str, family_id: str) -> bool:
         response = self.session.post(
             book_endpoint(self.box_name),
             data={
                 "id": class_id,
                 "day": target_day.strftime("%Y%m%d"),
                 "insist": 0,
-                "familyId": "",
+                "familyId": family_id,
```
Comment (pablobuenaposada) [inline src/client.py:80]: I am fine with this, even though I cannot really test in real life, but I don't see any unit test being modified which doesn't smell good

File: src/client.py [hunk 5]
```diff
@@ -45,32 +51,46 @@ def _login(email: str, password: str):
...
             },
         )
         if response.status_code == HTTPStatus.OK:
             response = response.json()
             if "bookState" in response and response["bookState"] == -2:
                 raise BookingFailed(MESSAGE_BOOKING_FAILED_NO_CREDIT)
+            if "bookState" in response and response["bookState"] == -12:
+                # self.logger.error(f"Booking unsuccesful. You cannot book the same session twice.")
```
Comment (pablobuenaposada) [inline src/client.py:88]: 🚓

File: src/main.py [hunk 6]
```diff
@@ -44,7 +61,8 @@ def main(email, password, booking_goals, box_name, box_id, days_in_advance):
      --password 1234
      --box-name lahuellacrossfit
      --box-id 3984
-     --booking-goal '{"0":{"time": "1815", "name": "Provenza"}}'
+     --booking-goal '{"0":{"time": "1815", "name": "Provenza"}}
```
Comment (pablobuenaposada) [inline src/main.py:64]: don't kill last `'` it's necessary

**Comment by pablobuenaposada** [comment]
check the build, is broken now:
https://github.com/pablobuenaposada/fitbot/actions/runs/6218682744/job/16875488721

File: Dockerfile (hunk 2 above)
Comment (jose-oc) [reply Dockerfile:9]: I prefer to have the application on a docker image without any other files than the ones needed to run the application. I've created a separate Dockerfile to run tests.

File: src/client.py (hunk 3 above)
Comment (jose-oc) [reply src/client.py:31]: That's because I like to have more info in plain English to identify issues quicker on the console :)

File: src/client.py (hunk 4 above)
Comment (jose-oc) [reply src/client.py:97]: I've tested in real world but you're right, I'll try and add some tests for this new parameter.

File: src/client.py [hunk 7]
```diff
@@ -45,32 +51,46 @@ def _login(email: str, password: str):
...
+                raise BookingFailed(response['errorMssg'])
```
[the remark this answers is not in the document]
Comment (jose-oc) [reply src/client.py]: Well, I thought it case a similar case and the message clarifies it but I'm fine with adding a new one

File: src/main.py [hunk 8]
```diff
@@ -1,40 +1,57 @@
...
-    except KeyError:  # did not found a matching booking goal
-        raise NoBookingGoal
+    except KeyError:  # There's nothing to book this day
+        raise NoClassOnTargetDayTime(f"There is no class to book on {day.strftime('%A, %Y-%m-%d')}. "
+                                     f"Either the time or the name could not be found in the input parameters.")
 
 
 def get_class_to_book(classes: list[dict], target_time: str, class_name: str):
+    if len(classes) == 0:
+        raise BoxClosed("Box is closed")
```
[the remark this answers is not in the document]
Comment (jose-oc) [reply src/main.py]: Done

File: src/main.py [hunk 9]
```diff
@@ -44,7 +61,8 @@ def main(email, password, booking_goals, box_name, box_id, days_in_advance):
      --password 1234
      --box-name lahuellacrossfit
      --box-id 3984
-     --booking-goal '{"0":{"time": "1815", "name": "Provenza"}}'
+     --booking-goal '{"0":{"time": "1815", "name": "Provenza"}}
+     --family_-id 123456'
```
[the remark this answers is not in the document]
Comment (jose-oc) [reply src/main.py:85]: Typo. Good spot

File: src/main.py (hunk 6 above)
Comment (jose-oc) [reply src/main.py:84]: Ups

**Comment by jose-oc** [comment]
> check the build, is broken now: https://github.com/pablobuenaposada/fitbot/actions/runs/6218682744/job/16875488721

it must be ok now

File: src/tests/test_client.py [hunk 10]
```diff
@@ -84,6 +82,7 @@ def test_get_classes(self, response, expected_classes):
             )
 
         with patch("requests.Session.get") as m_get:
+            m_get.return_value.status_code = HTTPStatus.OK
```
Comment (pablobuenaposada) [inline src/tests/test_client.py:85]: so the test passed without this so what's the point?

File: Dockerfile (hunk 2 above)
Comment (pablobuenaposada) [reply Dockerfile:9]: but then things like all the linter tools are not centralized in one file, honestly I don't see it

File: src/client.py (hunk 3 above)
Comment (pablobuenaposada) [reply src/client.py:31]: the exception we are using here can have a message on it, so I don't understand what more you need

**Comment by pablobuenaposada** [comment]
ok, the pull request went too far, I think you should change it so it add the family_id field and the upgrade of the dependencies and then we can discuss the rest in individual pull requests.
Please, also squash the commits into one.
Thanks for the work, I am willing to merge this but keeping the goal of every PR more closed.

**Comment by jose-oc** [comment]
I agree, there are different changes on this PR. I'm closing it and will create a new one.
